File: exercises/cpp/09_lr0_parser/lr0_parser.cpp
/*
 * ==============================================================================
 * Exercise 9: LR(0) Parser - Canonical Collection of Item Sets
 * ==============================================================================
 *
 * CONCEPT:
 * An LR(0) item is a production with a "dot" indicating the current parse
 * position. The canonical collection of LR(0) item sets forms the state
 * machine that drives bottom-up (shift-reduce) parsing. Each state encodes
 * what productions could be in progress given the symbols seen so far.
 *
 * The construction proceeds by repeatedly applying two operations:
 * - CLOSURE: expands an item set by adding items for non-terminals after dots.
 * - GOTO: advances the dot past a specific symbol, producing a new state.
 *
 * ALGORITHM:
 *   1. Augment grammar: S' -> S.
 *   2. I0 = CLOSURE({S' -> . S}).
 *   3. For each state I and each symbol X, compute GOTO(I, X).
 *      If the result is non-empty and new, add it to the collection.
 *   4. Repeat until no new states are generated.
 *
 * TIME COMPLEXITY:  O(|G|^2 * |V|) worst case.
 * SPACE COMPLEXITY: O(|G|^2) for storing all item sets.
 *
 * EXAMPLE INPUT:
 *   E -> E + T | T
 *   T -> T * F | F
 *   F -> ( E ) | id
 *
 * EXAMPLE OUTPUT:
 *   12 item sets (I0 through I11) with GOTO transitions.
 * ==============================================================================
 */

#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <algorithm>
#include <sstream>

/* --- Data structures --- */

struct Production {
    std::string lhs;
    std::vector<std::string> rhs;
};

struct LR0Item {
    int prod_index;  /* Index into the global productions vector */
    int dot_pos;     /* Position of the dot */

    bool operator<(const LR0Item &other) const {
        if (prod_index != other.prod_index)
            return prod_index < other.prod_index;
        return dot_pos < other.dot_pos;
    }

    bool operator==(const LR0Item &other) const {
        return prod_index == other.prod_index && dot_pos == other.dot_pos;
    }
};

/* An item set is a sorted set of items (using std::set for uniqueness) */
typedef std::set<LR0Item> ItemSet;

/* --- LR(0) Automaton Builder --- */

class LR0Automaton {
private:
    std::vector<Production> productions_;
    std::set<std::string> non_terminals_;
    std::vector<std::string> all_symbols_;
    std::string augmented_start_;

    std::vector<ItemSet> states_;
    std::map<ItemSet, int> state_index_;
    std::vector<std::tuple<int, std::string, int>> transitions_;

    bool is_non_terminal(const std::string &sym) const {
        return non_terminals_.count(sym) > 0;
    }

    const std::string *symbol_after_dot(const LR0Item &item) const {
        const Production &prod = productions_[item.prod_index];
        if (item.dot_pos >= static_cast<int>(prod.rhs.size()))
            return nullptr;
        return &prod.rhs[item.dot_pos];
    }

    ItemSet closure(const ItemSet &items) const {
        ItemSet result = items;
        bool changed = true;

        while (changed) {
            changed = false;
            ItemSet to_add;

            for (const auto &item : result) {
                const std::string *next_sym = symbol_after_dot(item);
                if (!next_sym || !is_non_terminal(*next_sym))
                    continue;

                for (int p = 0; p < static_cast<int>(productions_.size()); p++) {
                    if (productions_[p].lhs == *next_sym) {
                        LR0Item new_item = {p, 0};
                        if (result.find(new_item) == result.end()) {
                            to_add.insert(new_item);
                        }
                    }
                }
            }

            for (const auto &item : to_add) {
                if (result.insert(item).second)
                    changed = true;
            }
        }

        return result;
    }

    ItemSet goto_func(const ItemSet &items, const std::string &symbol) const {
        ItemSet kernel;

        for (const auto &item : items) {
            const std::string *next_sym = symbol_after_dot(item);
            if (next_sym && *next_sym == symbol) {
                LR0Item advanced = {item.prod_index, item.dot_pos + 1};
                kernel.insert(advanced);
            }
        }

        if (kernel.empty())
            return kernel;

        return closure(kernel);
    }

    void collect_symbols() {
        std::set<std::string> sym_set;
        for (const auto &prod : productions_) {
            sym_set.insert(prod.lhs);
            for (const auto &s : prod.rhs)
                sym_set.insert(s);
        }
        all_symbols_.assign(sym_set.begin(), sym_set.end());
    }

public:
    void add_non_terminal(const std::string &nt) {
        non_terminals_.insert(nt);
    }

    int add_production(const std::string &lhs,
                       const std::vector<std::string> &rhs) {
        non_terminals_.insert(lhs);
        productions_.push_back({lhs, rhs});
        return static_cast<int>(productions_.size()) - 1;
    }

    void build(const std::string &original_start) {
        augmented_start_ = original_start + "'";
        /* The augmented production must be at index 0 */
        /* It was already added by the caller */

        collect_symbols();

        /* I0 = closure({S' -> . S}) */
        ItemSet i0;
        i0.insert({0, 0});
        i0 = closure(i0);

        states_.push_back(i0);
        state_index_[i0] = 0;

        std::vector<int> worklist = {0};

        while (!worklist.empty()) {
            int state = worklist.front();
            worklist.erase(worklist.begin());

            for (const auto &sym : all_symbols_) {
                ItemSet goto_set = goto_func(states_[state], sym);
                if (goto_set.empty())
                    continue;

                int target;
                auto it = state_index_.find(goto_set);
                if (it != state_index_.end()) {
                    target = it->second;
                } else {
                    target = static_cast<int>(states_.size());
                    states_.push_back(goto_set);
                    state_index_[goto_set] = target;
                    worklist.push_back(target);
                }

                transitions_.push_back({state, sym, target});
            }
        }
    }

    void print_augmented_grammar() const {
        std::cout << "============================================================\n";
        std::cout << "  Augmented Grammar\n";
        std::cout << "============================================================\n";

        for (int i = 0; i < static_cast<int>(productions_.size()); i++) {
            std::cout << "  (" << i << ") " << productions_[i].lhs << " ->";
            for (const auto &s : productions_[i].rhs)
                std::cout << " " << s;
            std::cout << "\n";
        }
        std::cout << "============================================================\n";
    }

    void print_item(const LR0Item &item) const {
        const Production &prod = productions_[item.prod_index];
        std::cout << "    " << prod.lhs << " ->";

        for (int j = 0; j < static_cast<int>(prod.rhs.size()); j++) {
            if (j == item.dot_pos)
                std::cout << " .";
            std::cout << " " << prod.rhs[j];
        }
        if (item.dot_pos == static_cast<int>(prod.rhs.size()))
            std::cout << " .";
        std::cout << "\n";
    }

    void print_item_sets() const {
        std::cout << "\n============================================================\n";
        std::cout << "  Canonical Collection of LR(0) Item Sets\n";
        std::cout << "  Total states: " << states_.size() << "\n";
        std::cout << "============================================================\n";

        for (int i = 0; i < static_cast<int>(states_.size()); i++) {
            std::cout << "\n  I" << i << ":\n";
            std::cout << "  ----------------------------------------\n";
            for (const auto &item : states_[i]) {
                print_item(item);
            }
        }
        std::cout << "\n============================================================\n";
    }

    void print_transitions() const {
        std::cout << "\n============================================================\n";
        std::cout << "  GOTO Transitions\n";
        std::cout << "============================================================\n";

        for (const auto &t : transitions_) {
            std::cout << "  GOTO(I" << std::get<0>(t) << ", "
                      << std::get<1>(t) << ") = I" << std::get<2>(t) << "\n";
        }
        std::cout << "============================================================\n";
    }
};

int main() {
    std::cout << "============================================================\n";
    std::cout << "  LR(0) Parser - Canonical Collection Constructor\n";
    std::cout << "============================================================\n";

    LR0Automaton automaton;

    /* Register non-terminals */
    automaton.add_non_terminal("E'");
    automaton.add_non_terminal("E");
    automaton.add_non_terminal("T");
    automaton.add_non_terminal("F");

    /* Augmented grammar */
    automaton.add_production("E'", {"E"});         /* 0: E' -> E     */
    automaton.add_production("E",  {"E", "+", "T"}); /* 1: E -> E + T */
    automaton.add_production("E",  {"T"});           /* 2: E -> T     */
    automaton.add_production("T",  {"T", "*", "F"}); /* 3: T -> T * F */
    automaton.add_production("T",  {"F"});           /* 4: T -> F     */
    automaton.add_production("F",  {"(", "E", ")"}); /* 5: F -> ( E ) */
    automaton.add_production("F",  {"id"});          /* 6: F -> id    */

    automaton.print_augmented_grammar();
    automaton.build("E");
    automaton.print_item_sets();
    automaton.print_transitions();

    /* Second example */
    std::cout << "\n\n############################################################\n";
    std::cout << "  Second Example: S -> A A, A -> a A | b\n";
    std::cout << "############################################################\n";

    LR0Automaton automaton2;
    automaton2.add_non_terminal("S'");
    automaton2.add_non_terminal("S");
    automaton2.add_non_terminal("A");

    automaton2.add_production("S'", {"S"});
    automaton2.add_production("S",  {"A", "A"});
    automaton2.add_production("A",  {"a", "A"});
    automaton2.add_production("A",  {"b"});

    automaton2.print_augmented_grammar();
    automaton2.build("S");
    automaton2.print_item_sets();
    automaton2.print_transitions();

    return 0;
}

File: exercises/python/09_lr0_parser/lr0_parser.py
"""
==============================================================================
Exercise 9: LR(0) Parser - Canonical Collection of Item Sets
==============================================================================

CONCEPT:
An LR(0) item is a production with a dot (.) at some position in the
right-hand side, indicating how much of the production has been seen so far.
For example, from E -> E + T we get items:
  E -> . E + T    (nothing seen yet)
  E -> E . + T    (E has been seen)
  E -> E + . T    (E + has been seen)
  E -> E + T .    (complete item; ready to reduce)

The canonical collection of LR(0) item sets is the foundation for building
SLR(1), LALR(1), and canonical LR(1) parsers. Each item set represents a
parser state in the deterministic finite automaton used for shift-reduce
parsing.

ALGORITHM:
  1. Augment the grammar: add S' -> S (new start symbol).
  2. CLOSURE(I): For each item A -> alpha . B beta in I, add all items
     B -> . gamma for every production B -> gamma.
  3. GOTO(I, X): For each item A -> alpha . X beta in I, move the dot
     past X to get A -> alpha X . beta, then take the closure.
  4. Start with I0 = CLOSURE({S' -> . S}).
  5. For each item set and each grammar symbol, compute GOTO. If the
     result is a new set, add it to the collection. Repeat until no new
     sets are generated.

TIME COMPLEXITY:  O(|G|^2 * |V|) in the worst case, where |G| is grammar
                  size and |V| is the number of symbols.
SPACE COMPLEXITY: O(|G|^2) for storing all item sets.

EXAMPLE INPUT:
  E -> E + T | T
  T -> T * F | F
  F -> ( E ) | id

EXAMPLE OUTPUT:
  I0: { E' -> . E, E -> . E + T, E -> . T, T -> . T * F, ... }
  I1: { E' -> E ., E -> E . + T }
  ... (12 item sets total for this grammar)
==============================================================================
"""


class LR0Item:
    """
    Represents an LR(0) item: a production with a dot position.
    For production A -> X Y Z, dot_pos=1 means A -> X . Y Z.
    """

    def __init__(self, lhs, rhs, dot_pos=0):
        self.lhs = lhs
        self.rhs = tuple(rhs)
        self.dot_pos = dot_pos

    def is_complete(self):
        """The dot is at the end; this item is ready for reduction."""
        return self.dot_pos >= len(self.rhs)

    def symbol_after_dot(self):
        """Return the symbol immediately after the dot, or None if complete."""
        if self.is_complete():
            return None
        return self.rhs[self.dot_pos]

    def advance(self):
        """Return a new item with the dot moved one position to the right."""
        return LR0Item(self.lhs, self.rhs, self.dot_pos + 1)

    def __eq__(self, other):
        return (self.lhs == other.lhs and
                self.rhs == other.rhs and
                self.dot_pos == other.dot_pos)

    def __hash__(self):
        return hash((self.lhs, self.rhs, self.dot_pos))

    def __repr__(self):
        rhs_list = list(self.rhs)
        rhs_list.insert(self.dot_pos, ".")
        return f"{self.lhs} -> {' '.join(rhs_list)}"


class LR0Parser:
    """
    Builds the canonical collection of LR(0) item sets for a given grammar.
    """

    def __init__(self, grammar, start_symbol):
        """
        grammar: dict mapping non-terminal -> list of productions,
                 where each production is a list of symbols.
        start_symbol: the original start symbol (will be augmented).
        """
        self.original_start = start_symbol
        self.augmented_start = start_symbol + "'"

        # Augment the grammar
        self.grammar = {self.augmented_start: [[start_symbol]]}
        for lhs, prods in grammar.items():
            self.grammar[lhs] = [list(p) for p in prods]

        self.non_terminals = set(self.grammar.keys())
        self.terminals = set()
        for prods in self.grammar.values():
            for prod in prods:
                for sym in prod:
                    if sym not in self.non_terminals:
                        self.terminals.add(sym)

        self.all_symbols = self.non_terminals | self.terminals
        self.item_sets = []    # List of frozensets of LR0Items
        self.goto_table = {}   # (state_index, symbol) -> state_index

    def closure(self, items):
        """
        Compute the closure of a set of LR(0) items.
        For every item A -> alpha . B beta where B is a non-terminal,
        add B -> . gamma for all productions of B.
        """
        closure_set = set(items)
        worklist = list(items)

        while worklist:
            item = worklist.pop()
            next_sym = item.symbol_after_dot()

            if next_sym is not None and next_sym in self.non_terminals:
                for production in self.grammar[next_sym]:
                    new_item = LR0Item(next_sym, production, 0)
                    if new_item not in closure_set:
                        closure_set.add(new_item)
                        worklist.append(new_item)

        return frozenset(closure_set)

    def goto(self, item_set, symbol):
        """
        Compute GOTO(item_set, symbol).
        Move the dot past 'symbol' for all items where the dot precedes it,
        then take the closure.
        """
        moved_items = set()
        for item in item_set:
            if item.symbol_after_dot() == symbol:
                moved_items.add(item.advance())

        if not moved_items:
            return frozenset()

        return self.closure(moved_items)

    def build_canonical_collection(self):
        """
        Build the canonical collection of LR(0) item sets.
        Starts with I0 = closure({S' -> . S}) and iteratively computes
        GOTO for all symbols until no new item sets are found.
        """
        initial_item = LR0Item(self.augmented_start,
                               self.grammar[self.augmented_start][0], 0)
        start_set = self.closure({initial_item})
        self.item_sets = [start_set]

        # Map from frozenset -> index for quick lookup
        set_to_index = {start_set: 0}

        worklist = [0]
        while worklist:
            state_index = worklist.pop(0)
            current_set = self.item_sets[state_index]

            # Find all symbols that appear after a dot in this set
            symbols_after_dot = set()
            for item in current_set:
                sym = item.symbol_after_dot()
                if sym is not None:
                    symbols_after_dot.add(sym)

            for symbol in sorted(symbols_after_dot):
                goto_set = self.goto(current_set, symbol)
                if not goto_set:
                    continue

                if goto_set not in set_to_index:
                    new_index = len(self.item_sets)
                    self.item_sets.append(goto_set)
                    set_to_index[goto_set] = new_index
                    worklist.append(new_index)

                self.goto_table[(state_index, symbol)] = set_to_index[goto_set]

    def print_augmented_grammar(self):
        """Print the augmented grammar."""
        print("=" * 60)
        print("  Augmented Grammar")
        print("=" * 60)

        # Print augmented production first
        prod_number = 0
        for lhs in [self.augmented_start] + sorted(
                k for k in self.grammar if k != self.augmented_start):
            for rhs in self.grammar[lhs]:
                print(f"  ({prod_number}) {lhs} -> {' '.join(rhs)}")
                prod_number += 1
        print("=" * 60)

    def print_item_sets(self):
        """Print all item sets in the canonical collection."""
        print(f"\n{'=' * 60}")
        print(f"  Canonical Collection of LR(0) Item Sets")
        print(f"  Total states: {len(self.item_sets)}")
        print(f"{'=' * 60}")

        for idx, item_set in enumerate(self.item_sets):
            print(f"\n  I{idx}:")
            print(f"  {'-' * 40}")

            # Sort items for consistent output: augmented start first,
            # then by LHS, then by production, then by dot position
            sorted_items = sorted(
                item_set,
                key=lambda it: (
                    0 if it.lhs == self.augmented_start else 1,
                    it.lhs, it.rhs, it.dot_pos
                )
            )
            for item in sorted_items:
                print(f"    {item}")

        print(f"\n{'=' * 60}")

    def print_goto_table(self):
        """Print the GOTO transitions."""
        print(f"\n{'=' * 60}")
        print(f"  GOTO Transitions")
        print(f"{'=' * 60}")

        for (state, symbol), target in sorted(self.goto_table.items()):
            print(f"  GOTO(I{state}, {symbol}) = I{target}")

        print(f"{'=' * 60}")


def parse_grammar_text(grammar_text):
    """Parse a text-based grammar specification into a dictionary."""
    grammar = {}
    for line in grammar_text.strip().split("\n"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        lhs, rhs = line.split("->")
        lhs = lhs.strip()
        alternatives = rhs.split("|")
        productions = []
        for alt in alternatives:
            symbols = alt.strip().split()
            productions.append(symbols)
        if lhs in grammar:
            grammar[lhs].extend(productions)
        else:
            grammar[lhs] = productions
    return grammar


def main():
    print("=" * 60)
    print("  LR(0) Parser - Canonical Collection Constructor")
    print("=" * 60)

    grammar_text = """
        E -> E + T | T
        T -> T * F | F
        F -> ( E ) | id
    """

    grammar = parse_grammar_text(grammar_text)
    start_symbol = "E"

    parser = LR0Parser(grammar, start_symbol)
    parser.print_augmented_grammar()
    parser.build_canonical_collection()
    parser.print_item_sets()
    parser.print_goto_table()

    # Second smaller example
    print("\n\n" + "#" * 60)
    print("  Second Example: S -> A A, A -> a A | b")
    print("#" * 60)

    grammar_text_2 = """
        S -> A A
        A -> a A | b
    """

    grammar2 = parse_grammar_text(grammar_text_2)
    parser2 = LR0Parser(grammar2, "S")
    parser2.print_augmented_grammar()
    parser2.build_canonical_collection()
    parser2.print_item_sets()
    parser2.print_goto_table()


if __name__ == "__main__":
    main()
